Maker-Agents-Suit-Module/OSMAgent/src/big.queued.price.deviation.ts

```diff
@@ -0,0 +1,115 @@
+import {
+  Log,
+  TransactionEvent,
+  Trace,
+  HandleTransaction,
+  Finding,
+  FindingType,
+  FindingSeverity,
+} from "forta-agent";
+import Web3 from "web3";
+
+const PEEK_FUNCTION_SELECTOR = "0x59e02dd7";
+const LOG_VALUE_EVENT_SIGNATURE = "LogValue(bytes32)";
+
+export const createFinding = (contractAddress: string): Finding => {
+  return Finding.fromObject({
+    name: "MakerDAO OSM Contract Big Enqueued Price Deviation",
+    description:
+      "The new enqueued price deviate more than 6% from current price",
+    alertId: "MakerDAO-OSM-1",
+    type: FindingType.Suspicious,
+    severity: FindingSeverity.Info,
+    metadata: {
+      contractAddress: contractAddress,
+    },
+  });
+};
+
+const correctFunctionCalled = (trace: Trace): boolean => {
+  return trace.action.input === PEEK_FUNCTION_SELECTOR;
+};
+
+const callWasSuccessful = (trace: Trace): boolean => {
+  const results = new Web3().eth.abi.decodeParameters(
+    ["bytes32", "bool"],
+    //@ts-ignore
+    trace.result.output
+  );
+  return results[1];
+};
+
+const decodeNextValue = (trace: Trace): bigint => {
+  const results = new Web3().eth.abi.decodeParameters(
+    ["uint128", "bool"],
+    //@ts-ignore
+    trace.result.output
+  );
+  return BigInt(results[0]);
+};
+
+const getNextValuesForOSM = (contractAddress: string, traces: Trace[]) => {
+  return traces
+    .filter(
+      (trace) =>
+        trace.action.from === contractAddress &&
+        correctFunctionCalled(trace) &&
+        callWasSuccessful(trace)
+    )
+    .map(decodeNextValue);
+};
+
+const decodeCurrentValue = (log: Log): bigint => {
+  return BigInt(new Web3().eth.abi.decodeParameter("uint128", log.data) as any);
+};
+
+const getCurrentValues = (
+  contractAddress: string,
+  txEvent: TransactionEvent
+) => {
+  return txEvent
+    .filterEvent(LOG_VALUE_EVENT_SIGNATURE, contractAddress)
+    .map(decodeCurrentValue);
+};
+
+const abs = (a: bigint): bigint => {
+    return a < 0 ? -a : a;
+}
+
+const needToReport = (currentValue: bigint, nextValue: bigint): boolean => {
+    const bigDeviation: bigint = BigInt(6) * currentValue / BigInt(100);
+    return abs(currentValue - nextValue) > bigDeviation;
+}
+
+const checkOSMContract = (
+  contractAddress: string,
+  txEvent: TransactionEvent
+): Finding | null => {
+  const currentValues: bigint[] = getCurrentValues(contractAddress, txEvent);
+  const nextValues: bigint[] = getNextValuesForOSM(
+    contractAddress,
+    txEvent.traces
+  );
+
+  const lessLenght: number = Math.min(currentValues.length, nextValues.length);
+  for (let i = 0; i < lessLenght; i++) {
```

Comment:
May you elaborate this length topic. I'm just confused about lessLength and loop over it etc.

Reply:
I've already removed it. I thought I could have multiple lengths and in that case I wanted to iterate until one of the arrays were consumed. But both arrays must have the same length.